### PR TITLE
Km market token hoc

### PR DESCRIPTION
**Pull request by thirtyaughtsix**

### CODE REVIEW

**Comment by thirtyaughtsix** [comment]
Prolly need a squash merge rather than vanilla.

File: spec/markettoken/market-token.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      expect(parseInt(gasPrice)).toBeGreaterThan(0)
+    })
+
+    it('calls getPrivileged correctly', async () => {
+      const defaults = await token.getPrivileged({})
+      let tx = defaults[0]
+      let opts = defaults[1]
+      expect(tx).not.toBeNull(accounts[0])
+      expect(opts).not.toBeNull()
+      expect(Object.keys(opts).indexOf('gas')).toBeGreaterThan(-1)
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: i think a simple `gas in opts` is fine or a `opts.gas >= 121000` (note on that number. on the EVM 21000 gas is the least anything can cost. We pad every gas estimate by 100k. Thus, the gas returned for something unknown or unstated would be 121k)

File: spec/markettoken/market-token.spec.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      expect(parseInt(gasPrice)).toBeGreaterThan(0)
+    })
+
+    it('calls hasPrivilege correctly', async () => {
+      const defaults = await token.hasPrivilege("addr", {})
+      let tx = defaults[0]
+      let opts = defaults[1]
+      expect(tx).not.toBeNull(accounts[0])
+      expect(opts).not.toBeNull()
+      expect(Object.keys(opts).indexOf('gas')).toBeGreaterThan(-1)
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: same

File: spec/markettoken/market-token.spec.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      expect(Object.keys(opts).indexOf('from')).toBeGreaterThan(-1)
+      expect(Object.keys(opts).indexOf('gasPrice')).toBeGreaterThan(-1)
+      expect(opts['gas']).toBeGreaterThan(0)
+      let from = opts['from']
+      expect(from).not.toBeNull()
+      expect(from!.trim().length).toBeGreaterThan(0)
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: whats the `bang` for. is that a return-in-place kind of thing? i didn't know that...

**Comment by robrobbins** [comment]
also link to the FFA board ticket in the title and it will note it automagically...

File: spec/markettoken/market-token.spec.ts (hunk 3 above)
Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: yeah it's just like the swift `thingThatCanBeNull!.someMethOrParam()`

Comment (robrobbins) [reply spec/markettoken/market-token.spec.ts]: this ain't swift no more dorothy

File: spec/markettoken/market-token.spec.ts (hunk 1 above)
Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: updated in latest commit to test >= 121000

File: spec/markettoken/market-token.spec.ts (hunk 2 above)
Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: ditto

File: spec/markettoken/market-token.spec.ts (hunk 3 above)
Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: you ain't kidding...

File: spec/markettoken/market-token.spec.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    it('calls getPrivileged correctly', async () => {
+      const defaults = await token.getPrivileged({})
+      let tx = defaults[0]
+      let opts = defaults[1]
+      expect(tx).not.toBeNull(accounts[0])
+      expect(opts).not.toBeNull()
+      expect(Object.keys(opts).indexOf('gas')).toBeGreaterThan(-1)
+      expect(Object.keys(opts).indexOf('from')).toBeGreaterThan(-1)
+      expect(Object.keys(opts).indexOf('gasPrice')).toBeGreaterThan(-1)
+      expect(opts['gas']).toBeGreaterThanOrEqual(121000)
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: so this is correct.

the only time gas should be 121k is if there is no entry found in the `getGas` call. lots of things in the ABI gas estimations would come out less than 121k - we just default to it if we don't know anything else

File: spec/markettoken/market-token.spec.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    it('calls hasPrivilege correctly', async () => {
+      const defaults = await token.hasPrivilege("addr", {})
+      let tx = defaults[0]
+      let opts = defaults[1]
+      expect(tx).not.toBeNull(accounts[0])
+      expect(opts).not.toBeNull()
+      expect(Object.keys(opts).indexOf('gas')).toBeGreaterThan(-1)
+      expect(Object.keys(opts).indexOf('from')).toBeGreaterThan(-1)
+      expect(Object.keys(opts).indexOf('gasPrice')).toBeGreaterThan(-1)
+      expect(opts['gas']).toBeGreaterThanOrEqual(121000)
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: see above

File: spec/markettoken/market-token.spec.ts (hunk 4 above)
Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: updated the gas expectations... sounds like when i make chili....

File: spec/markettoken/market-token.spec.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    await token.at(w3, deployed.options.address)
+  })
+
+  describe('Class methods for Market Token', () => {
+
+    it('calls setPrivileged correctly', async () => {
+      const defaults = await token.setPrivileged("listing", "reserve", {})
+      let tx = defaults[0]
+      let opts = defaults[1]
+      let gas = token.getGas('listing')
```
Comment (robrobbins) [inline spec/markettoken/market-token.spec.ts]: wrong method yo

Comment (thirtyaughtsix) [reply spec/markettoken/market-token.spec.ts]: <sigh>

**Pull request merged**